specs/BrowserProcessExited.md

```diff
@@ -0,0 +1,186 @@
+# Background
+We have heard asks for a WebView2 API to easily track the `WebView2 Runtime`'s
+browser process exit. Manually waiting for the process to exit requires
+additional work on the host app, so we are proposing the `BrowserProcessExited`
+event. The `ProcessFailed` already lets app developers handle unexpected browser
+process exits for a WebView, this new API lets you listen to both expected and
+unexpected process termination from the `ICoreWebView2Environment` interface so
+you can, e.g., cleanup the user data folder when it's no longer in use. In this
+document we describe the new API. We'd appreciate your feedback.
+
+
+# Description
+The `BrowserProcessExited` event allows developers to subscribe event handlers
+to be run when the `WebView2 Runtime`'s browser process associated to a
+`CoreWebView2Environment` terminates. A key scenario is cleanup of the use data
+folder used by the `WebView2 Runtime`, which is locked while the runtime's
+browser process is active.
+
+This event is raised for both expected and unexpected browser process
+termination. The `ICoreWebView2BrowserProcessExitedEventArgs` interfaces lets
+app developers get the `BrowserProcessExitKind` so they can decide how to handle
+different exit kinds or bypass handling if an event handler for the
+`CoreWebView2`s `ProcessFailed` event (for
+`CoreWebView2ProcessFailedKind.BrowserProcessFailed`) is already registered.
+
+
+# Examples
+The following code snippets demonstrate how the `BrowserProcessExited` event can
+be used:
+
+## Win32 C++
+```cpp
+// Before closing the WebView, register a handler with code to run once the
+// browser process is terminated.
+EventRegistrationToken browserExitedEventToken;
+
+CHECK_FAILURE(m_webViewEnvironment->add_BrowserProcessExited(
+    Callback<ICoreWebView2BrowserProcessExitedEventHandler>(
+        [browserExitedEventToken, this](
+            ICoreWebView2Environment* sender,
+            ICoreWebView2BrowserProcessExitedEventArgs* args) {
+        COREWEBVIEW2_BROWSER_PROCESS_EXIT_KIND kind;
+        CHECK_FAILURE(args->get_BrowserProcessExitKind(&kind));
+
+        // Watch for graceful browser process exit. Let ProcessFailed event
+        // handler take care of failed browser process termination.
+        if (kind == COREWEBVIEW2_BROWSER_PROCESS_EXIT_KIND_NORMAL_EXIT)
+        {
+          CHECK_FAILURE(
+              m_webViewEnvironment->remove_BrowserProcessExited(browserExitedEventToken));
```

Comment:
There's no guarantee that there aren't more BrowserProcessExited events already queued up waiting to run the handler when remove_BrowserProcessExited is called.

Reply:
Lewis was looking at making this an async method instead of an event. An event seems like the more conventional option. But would an async method make this easier for the two expected scenarios (cleanup user data folder, and webview2 upgrade, are there other scenarios)?

Reply:
When the process exits, all registered handlers at that point will be invoked. After the event is raised, there won't be more events unless the app creates a new WebView.

The problem I have with an async method is there would be really no task/operation that gets started when calling the method, instead we "wait" for something to happen and that something could also happen more than once (if the app creates a new WebView after the process exits).

Reply:
I added an `Appendix` section for considering making this some sort of async method instead.